### ISSUE TITLE
[Question] memory_limit and n_jobs are getting ignored?

### ISSUE THREAD

**Issue by mfWeeWee** [automl/auto-sklearn#1221]
#Question
I am using auto-sklearn with a custom metric function. It seems that my memory_limit and n_jobs function are getting ignored. I have limit set 100*1024 and it doesn't go past 8 gb of ram and I have n_jobs = -1 and it only uses 1 thread. 
```
if __name__ == '__main__':  
  model = AutoSklearnClassifier(ensemble_size=100, #size of the end ensemble (minimum is 1)
                                                   time_left_for_this_task=120, #the number of seconds the process runs for
                                                   per_run_time_limit=5,
                                                   metric = custom_scorer #made with make_scorer from auto-sklearn
                                                   memory_limit = 100*1024,
                                                   n_jobs = -1,
                                                   resampling_strategy = 'cv',
                                                   resampling_strategy_arguments = {'folds': 5})

    model.fit(X_train, y_train) #begin fitting the search model
    print(model.sprint_statistics()) #print statistics for the search
    y_predictions = model.predict(X_test) #get predictions from the model
    acc = accuracy_score(y_test, y_predictions)
    print("Accuracy: %.3f" % acc)
    print(model.leaderboard())
    print(model.show_models())
```

System: 
 - Linux
 - Python 3.8
 - running through Anaconda
 - auto-sklearn 0.13.0

If I dont use __name__ part I can't use n_jobs it gives an error mentioned here #996 If I use _name__ part in the code, the code seems to stop after reading data, with no progress. 

Any help will be greatly appreciated

**Comment by eddiebergman** [comment]
Hi @mfWeeWee,

Unfortunatly, I can not reproduce the same problems on my end.

When running the following code I can see that AutoSklearn creates 4 jobs, uses as much memory as it needs and completes successfully (I do admit it's not very clear that it is running or give an indication of it's current progress).
```python
from autosklearn.classification import AutoSklearnClassifier
import autosklearn.pipeline.util as putil
from autosklearn.metrics import accuracy


if __name__ == '__main__':  
    X_train, y_train, X_test, y_test = putil.get_dataset('iris')

    model = AutoSklearnClassifier(ensemble_size=100, #size of the end ensemble (minimum is 1)
                                                   time_left_for_this_task=120, #the number of seconds the process runs for
                                                   per_run_time_limit=10,
                                                   metric = accuracy, #made with make_scorer from auto-sklearn
                                                   memory_limit = 3*1024,
                                                   n_jobs = 2,
                                                   resampling_strategy = 'cv',
                                                   resampling_strategy_arguments = {'folds': 5})

    print('fitting')
    model.fit(X_train, y_train) #begin fitting the search model
    print(model.sprint_statistics()) #print statistics for the search
    y_predictions = model.predict(X_test) #get predictions from the model
    acc = accuracy(y_test, y_predictions)
    print("Accuracy: %.3f" % acc)
    print(model.leaderboard())
    print(model.show_models())
```
* How large is your dataset?
* With `100*1024`, do you have 100 GB of RAM available? While this shouldn't matter as it works locally for me, it should definitely not need that much.
* How are checking how many jobs there are? With `top` I can see with `n_jobs = -1` gives me 4 jobs which is unfortunately all my tiny cpu can handle
     ```
     73525 skantify  20   0  881464 127768  20420 R  92.7   1.6   0:07.72 python                   
     73537 skantify  20   0  872596 118152  20052 R  80.4   1.5   0:02.42 python                   
     73538 skantify  20   0  939048 121372  22064 R  80.4   1.5   0:02.42 python                   
     73557 skantify  20   0  904308 123228  22484 R  69.8   1.5   0:02.10 python 
     ```

**Comment by mfWeeWee** [comment]
Hello @eddiebergman  Thank you for your answer. Yes I have a lot of memory and CPU power available. I check htop and I see that the code doesn't use as much as I think it should. For example TPOT used more than 50gb of ram (but also only 1 core) when evaluating the best classifier on this data. Data is consistent of 500 x 10000.

**Comment by eddiebergman** [comment]
Assuming you data consists of 64bit (8byte) floats in each slot then a brief calculation takes it as the dataset fits into 40MB of but it still doesn't reach anywhere near the RAM allocated. 

Sure the models will take some ram space as well but there's no reason it should use more memory than it needs. I don't know why TPOT consumes so much.

As for the job count, if TPOT is having the same issue, it could be something with the system you are running it on which unfortunately I can not help you with. The only pointer I can give is that autosklearn spawns processes but using a dask client such as [LocalCluster](https://docs.dask.org/en/latest/setup/single-distributed.html#localcluster) you can control how work is distributed. There is a parameter `dask_client` which overrides the default way of work distribution.

**Comment by eddiebergman** [comment]
Closing as there has been no feedback in the meantime. The RAM issue seems to be a misunderstanding, we will not use all the RAM available if there is no need. As for the CPU issue, we can not debug it from here or reproduce it. There are other CPU issues documented and hopefully fixing these will address this issue.

Closing this for now but if someone finds this and it is identical and not specified in any other issue, please feel free to comment and we will reopen this.
